src/ck-ldb/TreeBuilder.h

```diff
@@ -0,0 +1,348 @@
+    getPETopoTreeEdges(CkMyPe(), rootPE, NULL, -1, 4, &parent, &num_children, &children);
+    comm_parent[0] = parent;
+    if (num_children > 0)
+    {
+      comm_children[0].assign(children, children + num_children);
+      free(children);
+    }
+    comm_logic[0] = new MsgAggregator(num_children);
+
+    // root of tree (level 1)
+    if (CkMyPe() == rootPE)
+    {
+      RootLevel* level = new RootLevel();
+      level->configure(false, config["Root"]);
+      logic[1] = level;
+    }
+
+    if (CkMyPe() == 0 && !quietModeRequested)
+    {
+      auto& strategies = config["Root"]["strategies"];
+      if (strategies.size() == 1)
+      {
+        const std::string& strategy = strategies[0];
+        CkPrintf("[%d] TreeLB: Using PE_Root tree with strategy %s\n", CkMyPe(),
+                 strategy.c_str());
+      }
+      else
+      {
+        CkPrintf("[%d] TreeLB: Using PE_Root tree\n", CkMyPe());
+      }
+    }
+
+    return L;
+  }
+};
+
+class PE_Node_Root_Tree : public LBTreeBuilderCommon
+{
+ public:
+  virtual ~PE_Node_Root_Tree() {}
+
+  virtual uint8_t build(std::vector<LevelLogic*>& logic, std::vector<int>& comm_parent,
+                        std::vector<std::vector<int>>& comm_children,
+                        std::vector<LevelLogic*>& comm_logic, json& config)
+  {
+    uint8_t L = 3;  // num levels
+    reset(L, logic, comm_parent, comm_children, comm_logic);
+
+    int rootPE = 0;
+    if (config.contains("Root") && config["Root"].contains("pe"))
+    {
+      rootPE = config["Root"]["pe"];
+    }
+    CkAssert(rootPE >= 0 && rootPE < CkNumPes());
+
+    int mype = CkMyPe();
+    int mynode = CkMyNode();
+    int level1root = CkNodeFirst(mynode);
+    int level2root = rootPE;
+    LBManager* lbmgr = CProxy_LBManager(_lbmgr).ckLocalBranch();
+
+    // PE level (level 0)
+    int lvl = 0;
+    logic[lvl] = new PELevel(lbmgr);
+
+    // set up comm-tree between levels 0 and 1
+    if (mype == level1root)
+    {
+      // pes in my node excluding me
+      for (int pe = 0; pe < CkNumPes(); pe++)
+      {
+        if ((pe != mype) && (CkNodeOf(pe) == mynode)) comm_children[lvl].push_back(pe);
+      }
+    }
+    else
+    {
+      comm_parent[lvl] = level1root;
+    }
+
+    int step_freq_lvl2 = 1;
+    if (config.contains("Root") && config["Root"].contains("step_freq"))
+    {
+      step_freq_lvl2 = config["Root"]["step_freq"];
+    }
+
+    // node level (level 1)
+    lvl = 1;
+    if (mype == level1root)
+    {
+      std::vector<int> pes_in_node(CkNodeSize(mynode));
+      std::iota(pes_in_node.begin(), pes_in_node.end(), CkNodeFirst(mynode));
+
+      NodeLevel* level = new NodeLevel(lbmgr, pes_in_node);
+      level->configure(false, config["Process"], step_freq_lvl2);
+      logic[lvl] = level;
+
+      // set up comm-tree between levels 1 and 2
+      std::vector<int> level1_roots;
+      level1_roots.push_back(level2root);  // need this for getPETopoTreeEdges
+      for (int pe = 0; pe < CkNumPes(); pe++)
```

Comment:
I think we can replace this with a call to std::generate() to avoid a loop over all PEs